pandas/core/algorithms.py

```diff
@@ -112,16 +112,19 @@
 # --------------- #
 def _ensure_data(values: ArrayLike) -> np.ndarray:
     """
-    routine to ensure that our data is of the correct
-    input dtype for lower-level routines
+    Ensure values is of the correct input dtype for lower-level routines.
 
     This will coerce:
     - ints -> int64
     - uint -> uint64
-    - bool -> uint64 (TODO this should be uint8)
+    - bool -> uint8
     - datetimelike -> i8
     - datetime64tz -> i8 (in local tz)
     - categorical -> codes
+    - categorical[bool] without nulls -> uint8
+    - categorical[bool] with nulls -> ValueError: cannot convert float NaN to integer
```

Comment:
is this tested/intentional?

Reply:
looks like this was changed in #41256 although further investigation required on whether this is a latent bug/regression. Just updated the docstring for now to document the actual behavior.

Reply:
categorical is fast pathed in `mode` so does not pass through `_ensure_data`. So the regression fix in #42131 only required the `except TypeError` to fix.

In `duplicated` and `drop_duplicates` the categorical EA is passed through `_ensure_data` and so raises ValueError which is not caught by the fix in #42131.

So will need to change that but this is a regression from 1.2.5 so will need to be done separate so can be backported.

code sample based on test_drop_duplicates_categorical_bool

```python
import pandas as pd

print(pd.__version__)
tc = pd.Series(
    pd.Categorical(
        [True, False, True, False, pd.NA], categories=[True, False], ordered=True
    )
)
print(tc.duplicated())
```

```
1.2.5
0    False
1    False
2     True
3     True
4    False
dtype: bool
```

```python
1.3.4
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/tmp/ipykernel_47357/1277064552.py in <module>
      7     )
      8 )
----> 9 print(tc.duplicated())

~/miniconda3/envs/pandas-1.3.4/lib/python3.9/site-packages/pandas/core/series.py in duplicated(self, keep)
   2215         dtype: bool
   2216         """
-> 2217         res = self._duplicated(keep=keep)
   2218         result = self._constructor(res, index=self.index)
   2219         return result.__finalize__(self, method="duplicated")

~/miniconda3/envs/pandas-1.3.4/lib/python3.9/site-packages/pandas/core/base.py in _duplicated(self, keep)
   1230         self, keep: Literal["first", "last", False] = "first"
   1231     ) -> np.ndarray:
-> 1232         return duplicated(self._values, keep=keep)

~/miniconda3/envs/pandas-1.3.4/lib/python3.9/site-packages/pandas/core/algorithms.py in duplicated(values, keep)
    925     duplicated : ndarray[bool]
    926     """
--> 927     values, _ = _ensure_data(values)
    928     return htable.duplicated(values, keep=keep)
    929 

~/miniconda3/envs/pandas-1.3.4/lib/python3.9/site-packages/pandas/core/algorithms.py in _ensure_data(values)
    139             # i.e. all-bool Categorical, BooleanArray
    140             try:
--> 141                 return np.asarray(values).astype("uint8", copy=False), values.dtype
    142             except TypeError:
    143                 # GH#42107 we have pd.NAs present

ValueError: cannot convert float NaN to integer
```

Reply:
opened #44351 and will convert this to draft till fixed.